### PR TITLE
Optimize TPU Tokamax ring backward

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

**Problem:** Currently in the ring backward pass, each ring hop waits for the rotated dK/dV accumulators to arrive before starting the next hop's compute, so every hop cost compute + transfer. Additionally, every hop ran a full read, add, write pass over the dQ buffer, which is pure HBM traffic.

We introduce some optimizations

1/ We let the previous hop's dK/dV accumulators rotate while the current hop computes, so a hop costs max(compute, transfer) instead of (compute + transfer). 
2/ Currently each hop returns a reduced dQ and the outer loop added it into the full dQ buffer, a full read + write per hop. We pass the accumulator into the splash backward (`dq_carry_in`) as three partial buffers, so the kernel accumulates unreduced dQ partials in place, each partial buffer is read and written every third hop, and sum once after the scan. Only acctive when `dq_reduction_steps=3`.
3/ `ring_scan_unroll`: make the unroll configurable (was hardcoded to be off) for potential better scheduling.

The gain should grow with ring hop count because the optimizations reduce exposed comm and memory traffic from the backward loop.

# Tests

```
python3 -m pytest tests/unit/tokamax_ring_attention_test.py
```

```
python3 -m pytest tests/unit/configs_value_test.py -k tpu_tokamax_ring
```

```
python3 -m pytest tests/unit/tokamax_splash_attention_mask_test.py
```
TPU v5p (4 chips):
```
python3 -m pytest tests/unit/attention_test.py -k ring_context_parallel
```

all passed

## Benchmarks

On Llama-3.1-8B, 64 v5p chips

| | #4537 | This PR | MFU gain | Xprof |
|---|---|---|---|---|
| 256K, cp16, batch 2/shard | 177.3 TFLOP/s (38.6% MFU) | 185.8 (40.5%) | +1.9 pts ||
| 512K, cp16, batch 1/shard | 116.2 (25.3%) | 119.1 (25.9%) | +0.6 pts ||
| 1M, cp64, batch 2/shard  | 165.7 (36.1%) | 175.6 (38.3%) | +2.2 pts |[Xprof for base](http://xprof.corp.google.com/trace_viewer/htn-12015666495610921473),  [Xprof for opt](http://xprof.corp.google.com/trace_viewer/htn-5325066233409634149) , also see screenshot below|
| 1M, cp64, `dq_reduction_steps=0` | 128.8 (28.1%) | 129.5 (28.2%) | +0.1 pts ||

ring_scan_unroll | TFLOP/s/device | MFU
-- | -- | --
1 (default) | 175.6 | 38.3%
2 | 175.3 | 38.2%
4 | 168.8 | 36.8%
16 | 177.3 | 38.6%

For the 1M, cp64, batch 2/shard. Observe that the gap goes from 5.33 ms to 0.58 ms. At 1M cp64 that is 63 hops x 32 layers, about 9.6 s/step.

image
image

Repro

```
export LIBTPU_INIT_ARGS="--xla_tpu_scoped_vmem_limit_kib=65472 --xla_tpu_use_enhanced_launch_barrier=true --xla_tpu_enable_async_collective_fusion=true --xla_tpu_enable_async_collective_fusion_fuse_all_gather=true --xla_tpu_enable_async_collective_fusion_multiple_steps=true --xla_tpu_overlap_compute_collective_tc=true --xla_enable_async_all_gather=true --xla_enable_async_collective_permute=true --xla_tpu_enable_sparse_core_collective_offload_all_reduce=true --xla_tpu_enable_sparse_core_collective_offload_reduce_scatter=true --xla_tpu_enable_sparse_core_collective_offload_3d_all_gather=true"

python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml \
  run_name=<name> base_output_directory=<gcs path> \
  model_name=llama3.1-8b dataset_type=synthetic enable_checkpointing=false \
  attention=flash use_tokamax_splash=true use_jax_splash=false \
  context_parallel_strategy=ring context_parallel_load_balance=true packing=false \
  dq_reduction_steps=3 remat_policy=full \
  sa_block_q=2048 sa_block_kv=2048 sa_block_q_dkv=2048 sa_block_kv_dkv=2048 \
  ici_tensor_parallelism=1 ici_context_parallelism=<cp> ici_fsdp_parallelism=<fsdp> \
  per_device_batch_size=<pdbs> max_target_length=<context> steps=10 \
  profiler=xplane profiler_steps=3 skip_first_n_steps_for_profiler=3
```

|  | S | CP | FSDP | pdbs |
|---|---|---|---|---|
| 256K | 262144 | 16 | 4 | 0.125 | 
| 512K | 524288 | 16 | 4 | 0.0625 |
| 1M  | 1048576 | 64 | 1 | 0.03125 | 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -745,6 +745,7 @@ class SplashAttention(BaseModel):
   sa_k_layout: str = Field("HEAD_DIM_MINOR", description="Layout for K in splash attention.")
   sa_v_layout: str = Field("HEAD_DIM_MINOR", description="Layout for V in splash attention.")
   use_splash_scheduler: bool = Field(False, description="Use experimental splash attention scheduler.")
+  ring_scan_unroll: NonNegativeInt = Field(1, description="Unroll factor for the Tokamax ring attention scan.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py]: let's disable this flag (force to be 1) when ring is not used

File: src/maxtext/kernels/attention/tokamax_ring_attention.py [hunk 2]
```diff
@@ -252,6 +252,9 @@ def build_splash_config(
         "TPU Tokamax ring attention with context_parallel_load_balance=True requires "
         f"sa_block_q_dkv ({block_q_dkv}) to be a multiple of {tokamax_splash_kernel.NUM_LANES} after clamping."
     )
+  # Ring uses the dynamic-grid dKV path, so mirror Splash's small-kv_steps guard here.
+  if dq_reduction_steps == 3 and kv_seq_len_per_shard // block_kv_dkv <= 3:
+    dq_reduction_steps = 0
```
Comment (NuojCheng) [inline src/maxtext/kernels/attention/tokamax_ring_attention.py:257]: make sense

File: src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py [hunk 3]
```diff
@@ -211,32 +212,27 @@ def _ring_attention_bwd(
       axis_name=ring_axis,
       perm=[(i, (i + 1) % ring_axis_size) for i in range(ring_axis_size)],
   )
-  dq_accum = jnp.zeros(q.shape, dtype=jnp.float32)
+  use_dq_carry = config.dq_reduction_steps == 3
+  dq_accum = jnp.zeros((3, *q.shape), dtype=jnp.float32) if use_dq_carry else jnp.zeros(q.shape, dtype=jnp.float32)
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py:216]: more HBM memory usage?

File: src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py [hunk 4]
```diff
@@ -259,30 +255,72 @@ def body(carry, i: int):
...
+        k_current,
+        v_current,
+        segment_ids_current,
+        _,
+    ) = carry
+    dk_next = shift(dk_accum + dk_pending.astype(jnp.float32))
+    dv_next = shift(dv_accum + dv_pending.astype(jnp.float32))
+    k_next, v_next, segment_ids_next = rotate_kv(k_current, v_current, segment_ids_current)
+    dq_i, dk_i, dv_i, dsinks = compute_step(i, k_current, v_current, segment_ids_current, dq_accum)
+    dq_accum = dq_i
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py:289]: I like the changes above. It looks more clean now, great job!

File: src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py [hunk 5]
```diff
@@ -363,6 +401,10 @@ def _ring_attention_fwd(
       ring_axis=ring_axis,
       expected_ring_size=expected_ring_size,
   )
+  # The ring backward reads the merged residuals, not the per-hop splash outputs.
+  if config.residual_checkpoint_name is not None:
+    out = ad_checkpoint.checkpoint_name(out, name=config.residual_checkpoint_name)
+    logsumexp = ad_checkpoint.checkpoint_name(logsumexp, name=config.residual_checkpoint_name)
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py:407]: I would expect a test to check whether the remat policy works here, either a unit test or a test in PR description.

**Review by NuojCheng — COMMENTED** [review]
Great PR! I am exciting to see more optimizations on ring attention. For the experiments comparing perf before/after this PR, could you also provide the losses and see if they match? One example should be fine.

File: src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py (hunk 3 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py:216]: Yes

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: it's already default to be 1 and non-ring attentions never pass anything to change it. should we be extra defensive still?

Comment (NuojCheng) [reply src/maxtext/configs/types.py]: yes, we should try to scream out if a flag is passed but not used

**Comment by huytransformer** [comment]
> Great PR! I am exciting to see more optimizations on ring attention. For the experiments comparing perf before/after this PR, could you also provide the losses and see if they match? One example should be fine.

here is the abbrev logs for 1M run pair so u can see the matched loss at every step:

```
# Run 1 (PR2RB_1m_base)
I0729 19:52:57.134704 140516855904064 metric_logger.py:226] completed step: 0, seconds: 3.456, TFLOP/s/device: 8245.657, Tokens/s/device: 9481.454, total_weights: 2097152, loss: 12.262, lm_loss: 12.262, perplexity: 211569.500
I0729 19:55:49.091287 140516855904064 metric_logger.py:226] completed step: 1, seconds: 0.379, TFLOP/s/device: 75136.046, Tokens/s/device: 86396.870, total_weights: 2097152, loss: 12.262, lm_loss: 12.262, perplexity: 211569.500
I0729 19:58:41.092984 140516855904064 metric_logger.py:226] completed step: 2, seconds: 172.700, TFLOP/s/device: 165.009, Tokens/s/device: 189.740, total_weights: 2097152, loss: 11.937, lm_loss: 11.937, perplexity: 152810.750
I0729 20:01:33.073593 140516855904064 metric_logger.py:226] completed step: 3, seconds: 343.946, TFLOP/s/device: 82.853, Tokens/s/device: 95.271, total_weights: 2097152, loss: 11.568, lm_loss: 11.568, perplexity: 105649.188
I0729 20:07:56.151026 140516855904064 metric_logger.py:226] completed step: 4, seconds: 0.009, TFLOP/s/device: 3095825.491, Tokens/s/device: 3559804.454, total_weights: 2097152, loss: 11.265, lm_loss: 11.265, perplexity: 78021.938
I0729 20:07:56.160824 140516855904064 metric_logger.py:226] completed step: 5, seconds: 171.982, TFLOP/s/device: 165.698, Tokens/s/device: 190.532, total_weights: 2097152, loss: 11.002, lm_loss: 11.002, perplexity: 59995.281
I0729 20:10:49.653540 140516855904064 metric_logger.py:226] completed step: 6, seconds: 383.077, TFLOP/s/device: 74.390, Tokens/s/device: 85.539, total_weights: 2097152, loss: 10.814, lm_loss: 10.814, perplexity: 49715.031
I0729 20:13:41.627545 140516855904064 metric_logger.py:226] completed step: 7, seconds: 1.512, TFLOP/s/device: 18847.894, Tokens/s/device: 21672.674, total_weights: 2097152, loss: 10.676, lm_loss: 10.676, perplexity: 43321.688
I0729 20:16:33.625483 140516855904064 metric_logger.py:226] completed step: 8, seconds: 171.990, TFLOP/s/device: 165.691, Tokens/s/device: 190.523, total_weights: 2097152, loss: 10.596, lm_loss: 10.596, perplexity: 39985.781
I0729 20:19:25.592107 140516855904064 metric_logger.py:226] completed step: 9, seconds: 171.976, TFLOP/s/device: 165.704, Tokens/s/device: 190.539, total_weights: 2097152, loss: 10.549, lm_loss: 10.549, perplexity: 38132.312

# Run 2 (PR2RB_1m_opt)
I0729 19:56:18.861211 134912273610560 metric_logger.py:226] completed step: 0, seconds: 3.585, TFLOP/s/device: 7949.116, Tokens/s/device: 9140.470, total_weights: 2097152, loss: 12.262, lm_loss: 12.262, perplexity: 211569.500
I0729 19:59:01.092090 134912273610560 metric_logger.py:226] completed step: 1, seconds: 0.374, TFLOP/s/device: 76267.369, Tokens/s/device: 87697.747, total_weights: 2097152, loss: 12.262, lm_loss: 12.262, perplexity: 211569.500
I0729 20:01:43.482037 134912273610560 metric_logger.py:226] completed step: 2, seconds: 162.277, TFLOP/s/device: 175.608, Tokens/s/device: 201.927, total_weights: 2097152, loss: 11.937, lm_loss: 11.937, perplexity: 152817.000
I0729 20:04:25.731209 134912273610560 metric_logger.py:226] completed step: 3, seconds: 324.609, TFLOP/s/device: 87.789, Tokens/s/device: 100.946, total_weights: 2097152, loss: 11.568, lm_loss: 11.568, perplexity: 105638.438
I0729 20:10:30.671063 134912273610560 metric_logger.py:226] completed step: 4, seconds: 0.039, TFLOP/s/device: 733005.984, Tokens/s/device: 842863.390, total_weights: 2097152, loss: 11.265, lm_loss: 11.265, perplexity: 78020.688
I0729 20:10:30.680328 134912273610560 metric_logger.py:226] completed step: 5, seconds: 162.221, TFLOP/s/device: 175.669, Tokens/s/device: 201.996, total_weights: 2097152, loss: 11.002, lm_loss: 11.002, perplexity: 59992.281
I0729 20:13:15.073705 134912273610560 metric_logger.py:226] completed step: 6, seconds: 364.939, TFLOP/s/device: 78.087, Tokens/s/device: 89.790, total_weights: 2097152, loss: 10.814, lm_loss: 10.814, perplexity: 49719.031
I0729 20:15:57.304623 134912273610560 metric_logger.py:226] completed step: 7, seconds: 2.122, TFLOP/s/device: 13426.714, Tokens/s/device: 15439.009, total_weights: 2097152, loss: 10.676, lm_loss: 10.676, perplexity: 43321.875
I0729 20:18:39.557860 134912273610560 metric_logger.py:226] completed step: 8, seconds: 162.279, TFLOP/s/device: 175.605, Tokens/s/device: 201.923, total_weights: 2097152, loss: 10.596, lm_loss: 10.596, perplexity: 39984.062
I0729 20:21:21.774135 134912273610560 metric_logger.py:226] completed step: 9, seconds: 162.231, TFLOP/s/device: 175.658, Tokens/s/device: 201.984, total_weights: 2097152, loss: 10.549, lm_loss: 10.549, perplexity: 38131.625
```

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: Done!

File: src/maxtext/kernels/attention/tokamax_ring_attention.py (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/kernels/attention/tokamax_ring_attention.py:257]: is this "3" a magic number? Or what's the reason to use as condition?

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 6]
```diff
@@ -1726,6 +1729,8 @@ def _bwd_cost_estimate(
         interpret=config.interpret,
         metadata=metadata,
     )(*args, dq, dk, dv)
+  if return_unreduced_dq and dq_reduction_steps == 3:
+    return dq_unreduced, dk.astype(jnp.float32), dv.astype(jnp.float32)
```
Comment (RissyRan) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1733]: Do you need f32 in this case? I see dk, and dv are aligned. Or `q.dtype`?

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py]: Is there any constraints for this config? i.e. max value?

**Review by RissyRan — COMMENTED** [review]
LGTM, just a few minor comment!

Also, recommend to push to maxtext directly as writer. So Gemini PR review can be triggered for 1st round review.

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: the only hard constraint is non-negative int. 0 or >= CP means full unroll. I think this was a sensible choice because the max value isn't immediately available here and would only complicate the code path. This is not convoluted and there's no error to prevent...

File: src/maxtext/kernels/attention/tokamax_ring_attention.py (hunk 2 above)
Comment (huytransformer) [reply src/maxtext/kernels/attention/tokamax_ring_attention.py:257]: great question! 3 kind of is a magic number.... but there's reason behind it. TLDR: instead of keeping 1 buffer per kv block (which grows linearly with # blocks) we allocate & reuse 3. We need to read/add/write so that's the smallest safe value. Plus if u have <=3 KV blocks then there's really nothing to reuse.

File: src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py (hunk 5 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/ring_attention_kernel.py:407]: done!

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py (hunk 6 above)
Comment (huytransformer) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:1733]: ya. we usually require f32 for accumulators to avoid errors (ring always requests fp32 grads and dk/dv are actually already f32 here)

**Comment by huytransformer** [comment]
> LGTM, just a few minor comment!
> 
> Also, recommend to push to maxtext directly as writer. So Gemini PR review can be triggered for 1st round review.

TY I thought merging from my own fork was better practice. Kept wondering why gemini review didn't work for my PRs...

**Review by RissyRan — APPROVED** [review]

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (NuojCheng) [reply src/maxtext/configs/types.py]: I believe this number cannot be larger than scan length, not sure what error gonna get though

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: good catch but full unroll, no error
